### PR TITLE
Lookup Boundary

### LINKED ISSUE #115381: InheritanceBoundary
The big advantage of implementing multiple views with a single widget tree and the OverlayPortal proposal is that overlays/views have access to the surrounding inherited widgets at their anchor point in the widget tree. However, some inherited widgets should not flow into a View or OverlayPortal. 

Examples of widgets that should _not_ be available:
* the Material widget - a view needs to contain its own Material for buttons to draw splash effects on, it should not draw the splashes in the parent view
* the view metrics from MediaQuery (needs to be split into ViewQuery and PlatformQuery, see https://github.com/flutter/flutter/issues/115375)
* Navigator - it will be confusing if a sub-view can modify the navigation stack of the parent view, if that's required, the Navigator should be manually injected
* Overlay - subviews need to contain their own Overlay if they want to show something in an overlay. It would be confusing if the subview's text handles and toolbars are drawn into the parent view's overlay

To support this, we could introduce an InheritanceBoundary that View/OverlayPortal could use to stop certain tagged InheritedWidget from propagating to their children. The InheritedWidget itself gets to decide whether it wants to propagate through the boundary or not.

Implementation Idea:

InheritanceBoundary is a mixin on Element that overrides `Element._updateInheritance`. Since that method is private (and should likely remain so) it needs to be implemented in framework.dart. Instead of just copying the list of inherited widgets from its parent, InheritanceBoundary will only copy the ones that are allowed to flow through the boundary. The `InheritedWidget` constructor is extended with a `stopsAtInheritanceBoundary` argument, which defaults to false. For an InheritedWidget subclass, it must be hard-coded to true or false and is not allowed to change. This property is used by the InheritanceBoundary to determine whether a given inherited widget is supposed to flow through the boundary or not.

### PR DESCRIPTION
**Pull request by goderbauer**
Fixes https://github.com/flutter/flutter/issues/115381.

## Design Discussion

This change introduces a `stopAtLookUpBoundary` parameter to the various lookup methods on `BuildContext` and offers a consistent API between all lookup methods, which I really like. However, it comes at the cost of adding a field (`_lookUpBoundary`) to every `Element`. This could probably be slightly optimized by replacing the `_inheritedWidgets` with an immutable data object that incorporates the hashmap of inherited elements and the lookUpBoundary field. That way, most Elements would just point to the same data object and not need storage for an extra field. Only Elements of InheritedWidgets and Elements who are actually lookup boundaries would have slightly increased memory needs because they will have to create a new data object with that extra field for the lookup boundary.

An alternative approach would forego a common API among all lookup methods. It would keep the `stopAtLookUpBoundary` parameter on the various `find*` lookup methods, but remove it from the `dependOn*` and `get*` lookup methods. Instead of deciding at lookup time whether to retrieve InheritedWidgets past the lookup boundary, the InheritedWidgets themselves would decide whether they want to be visible through the lookup boundary. This would be controlled via a constructor argument to InheritedWidget or an overridable getter. At a lookup boundary, we would filter the hashmap of inherited widgets to remove those who have specified that they don't want to be visible through the lookup boundary. This would make adding and moving lookup boundaries more expensive because it requires filtering a potentially long list of inherited widgets and the lookup APIs on BuildContext would be less consistent. But it would remove the extra check from the lookup implementation for inherited widgets.

| Approach in PR      | Alternative Approach |
| ----------- | ----------- |
| + consistent API for all lookup methods      | - `find` methods have different API from `depend`/`get` methods       |
| - requires extra memory (1 field per element, can be optimized to one per InheritedWidget/LookupBoundary)   | + no extra memory        |
| - requires extra check for lookup of inherited widgets | + no extra check for inherited widget lookup |
| - each element needs to copy lookupboundary reference from parent in mount/activate | - adding/moving lookup boundaries is slower due to filtering |
| + flexible (at lookup it is decided whether an InheritedWidget should be seen) | - less flexible (InheritedWidget decides wether it wants to be seen) |
| | - requires extending the `PersistentHashMap` interface to allow for iteration for filtering |

### CODE REVIEW

**Comment by goderbauer** [comment]
@gspencergoog @LongCatIsLooong @Hixie Curious to hear what you think of this approach. I am a little torn between the two alternatives discussed in the PR description. I am currently leaning towards having consistent APIs as implemented in this PR, though. Let me know if you have any other ideas that I might have missed.

File: packages/flutter/lib/src/widgets/framework.dart [hunk 1]
```diff
@@ -2262,7 +2264,15 @@ abstract class BuildContext {
   /// [InheritedWidget] subclasses that supports partial updates, like
   /// [InheritedModel]. It specifies what "aspect" of the inherited
   /// widget this context depends on.
-  T? dependOnInheritedWidgetOfExactType<T extends InheritedWidget>({ Object? aspect });
+  ///
+  /// When `stopAtLookUpBoundary` is set to true, null is returned if
+  /// no inherited widget of the specified type exists between this
+  /// [BuildContext] and the closest ancestor lookup boundary.
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart:2270]: reading this i would just ask myself "what is a lookup boundary". I think every mention of lookup boundary should link to `Element.isLookUpBoundary`.

**Comment by Hixie** [comment]
One con of this approach is that it is more bug-prone in the sense that if there are multiple ways to look up a widget, some might set the flag and some not. Given that we usually wrap calls into `.of` methods etc, that's probably rare, though.

**Comment by goderbauer** [comment]
Adding the `stopAtLookUpBoundary` to `getElementForInheritedWidgetOfExactType ` is breaking `package:provider`, which is overriding that method in a custom element implementation: https://github.com/rrousselGit/provider/blob/bd938bdd7a32692c489a74ae2fd31fe57f59c6c8/lib/src/inherited_provider.dart#L381

**Comment by Hixie** [comment]
@Piinks was saying the other day that there may be a way to add parameters using `dart fix`?

**Comment by goderbauer** [comment]
> there may be a way to add parameters using dart fix?

We would still have to fix the logic inside the body to account for the parameter, which likely cannot be done automatically.

Unfortunately, package:provider is forwarding the call to a different Element inside the body, which means we can't fix them to be compatible with both master and stable...

**Comment by Piinks** [comment]
Also, (sorry) dart fix does not apply to overridden subclass methods. It only applies to the specific class the fix has been defined for. There is a feature request for applying to subclasses though: https://github.com/dart-lang/sdk/issues/36027

**Comment by goderbauer** [comment]
Closing this in favor of https://github.com/flutter/flutter/pull/116429.

**Pull request closed without merging**